File: clients/client-python/gravitino/dto/rel/column_dto.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional, Union, cast

from dataclasses_json import DataClassJsonMixin, config

from gravitino.api.column import Column
from gravitino.api.expressions.expression import Expression
from gravitino.api.types.json_serdes.type_serdes import TypeSerdes
from gravitino.api.types.type import Type
from gravitino.api.types.types import Types
from gravitino.dto.rel.expressions.json_serdes.column_default_value_serdes import (
    ColumnDefaultValueSerdes,
)
from gravitino.dto.rel.expressions.literal_dto import LiteralDTO
from gravitino.utils.precondition import Precondition


@dataclass
class ColumnDTO(Column, DataClassJsonMixin):
    """Represents a Model DTO (Data Transfer Object)."""

    _name: str = field(metadata=config(field_name="name"))
    """The name of the column."""

    _data_type: Type = field(
        metadata=config(
            field_name="type",
            encoder=TypeSerdes.serialize,
            decoder=TypeSerdes.deserialize,
        )
    )
    """The data type of the column."""

    _comment: str = field(metadata=config(field_name="comment"))
    """The comment associated with the column."""

    _default_value: Optional[Union[Expression, List[Expression]]] = field(
        default_factory=lambda: Column.DEFAULT_VALUE_NOT_SET,
        metadata=config(
            field_name="defaultValue",
            encoder=ColumnDefaultValueSerdes.serialize,
            decoder=ColumnDefaultValueSerdes.deserialize,
            exclude=lambda value: value is None
            or value is Column.DEFAULT_VALUE_NOT_SET,
        ),
    )
    """The default value of the column."""

    _nullable: bool = field(default=True, metadata=config(field_name="nullable"))
    """Whether the column value can be null."""

    _auto_increment: bool = field(
        default=False, metadata=config(field_name="autoIncrement")
    )
    """Whether the column is an auto-increment column."""

    def name(self) -> str:
        return self._name

    def data_type(self) -> Type:
        return self._data_type

    def comment(self) -> str:
        return self._comment

    def nullable(self) -> bool:
        return self._nullable

    def auto_increment(self) -> bool:
        return self._auto_increment

    def default_value(self) -> Union[Expression, List[Expression]]:
        return self._default_value

    def validate(self) -> None:
        Precondition.check_string_not_empty(
            self._name, "Column name cannot be null or empty."
        )
        Precondition.check_argument(
            self._data_type is not None, "Column data type cannot be null."
        )
        non_nullable_condition = (
            not self._nullable
            and isinstance(self._default_value, LiteralDTO)
            and cast(LiteralDTO, self._default_value).data_type()
            == Types.NullType.get()
        )
        Precondition.check_argument(
            not non_nullable_condition,
            f"Column cannot be non-nullable with a null default value: {self._name}.",
        )

    @staticmethod
    def builder() -> ColumnDTO.Builder:
        return ColumnDTO.Builder()

    def __eq__(self, other: ColumnDTO) -> bool:
        if not isinstance(other, ColumnDTO):
            return False
        return (
            self._name == other._name
            and self._data_type == other._data_type
            and self._comment == other._comment
            and self._nullable == other._nullable
            and self._auto_increment == other._auto_increment
            and self._default_value == other._default_value
        )

    def __hash__(self) -> int:
        return hash(
            (
                self._name,
                self._data_type,
                self._comment,
                self._nullable,
                self._auto_increment,
                (
                    None
                    if self._default_value is Column.DEFAULT_VALUE_NOT_SET
                    else self._default_value
                ),
            )
        )

    class Builder:
        def __init__(self) -> None:
            self._name = None
            self._data_type = None
            self._comment = None
            self._nullable: bool = True
            self._auto_increment: bool = False
            self._default_value = None

        def with_name(self, name: str) -> ColumnDTO.Builder:
            """Sets the name of the column.

            Args:
                name (str): The name of the column.

            Returns:
                ColumnDTO.Builder: The builder instance.
            """
            self._name = name
            return self

        def with_data_type(self, data_type: Type) -> ColumnDTO.Builder:
            """Sets the data type of the column.

            Args:
                data_type (Type): The data type of the column.

            Returns:
                ColumnDTO.Builder: The builder instance.
            """
            self._data_type = data_type
            return self

        def with_comment(self, comment: str) -> ColumnDTO.Builder:
            """Sets the comment associated with the column.

            Args:
                comment (str): The comment associated with the column.

            Returns:
                ColumnDTO.Builder: The builder instance.
            """
            self._comment = comment
            return self

        def with_nullable(self, nullable: bool) -> ColumnDTO.Builder:
            """Sets whether the column value can be null.

            Args:
                nullable (bool): Whether the column value can be null.

            Returns:
                ColumnDTO.Builder: The builder instance.
            """
            self._nullable = nullable
            return self

        def with_auto_increment(self, auto_increment: bool) -> ColumnDTO.Builder:
            """Sets whether the column is an auto-increment column.

            Args:
                auto_increment (bool): Whether the column is an auto-increment column.

            Returns:
                ColumnDTO.Builder: The builder instance.
            """
            self._auto_increment = auto_increment
            return self

        def with_default_value(self, default_value: Expression) -> ColumnDTO.Builder:
            """Sets the default value of the column.

            Args:
                default_value (Expression): The default value of the column.

            Returns:
                ColumnDTO.Builder: The builder instance.
            """
            self._default_value = default_value
            return self

        def build(self) -> ColumnDTO:
            """Builds a Column DTO based on the provided builder parameters.

            Returns:
                ColumnDTO: A new ColumnDTO instance.

            Raises:
                IllegalArgumentException: If required, fields name and data type are not set.
            """
            Precondition.check_argument(
                self._name is not None, "Column name cannot be null"
            )
            Precondition.check_argument(
                self._data_type is not None, "Column data type cannot be null"
            )
            return ColumnDTO(
                _name=self._name,
                _data_type=self._data_type,
                _comment=self._comment,
                _nullable=self._nullable,
                _auto_increment=self._auto_increment,
                _default_value=(
                    Column.DEFAULT_VALUE_NOT_SET
                    if self._default_value is None
                    else self._default_value
                ),
            )

File: common/src/main/java/org/apache/gravitino/dto/rel/ColumnDTO.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.dto.rel;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.google.common.base.Preconditions;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import org.apache.gravitino.Audit;
import org.apache.gravitino.dto.AuditDTO;
import org.apache.gravitino.dto.rel.expressions.LiteralDTO;
import org.apache.gravitino.json.JsonUtils;
import org.apache.gravitino.rel.Column;
import org.apache.gravitino.rel.expressions.Expression;
import org.apache.gravitino.rel.types.Type;
import org.apache.gravitino.rel.types.Types;

/** Represents a Column DTO (Data Transfer Object). */
@EqualsAndHashCode
@ToString
public class ColumnDTO implements Column {

  @JsonProperty("name")
  private String name;

  @JsonProperty("type")
  @JsonSerialize(using = JsonUtils.TypeSerializer.class)
  @JsonDeserialize(using = JsonUtils.TypeDeserializer.class)
  private Type dataType;

  @JsonProperty("comment")
  private String comment;

  @JsonProperty("nullable")
  private boolean nullable = true;

  @JsonProperty("autoIncrement")
  private boolean autoIncrement = false;

  @JsonProperty("defaultValue")
  // the NON_EMPTY annotation is used to avoid serializing the defaultValue field if it is not set
  // or null.
  @JsonInclude(JsonInclude.Include.NON_EMPTY)
  @JsonSerialize(using = JsonUtils.ColumnDefaultValueSerializer.class)
  @JsonDeserialize(using = JsonUtils.ColumnDefaultValueDeserializer.class)
  private Expression defaultValue = Column.DEFAULT_VALUE_NOT_SET;

  private ColumnDTO() {}

  /**
   * Constructs a Column DTO.
   *
   * @param name The name of the column.
   * @param dataType The data type of the column.
   * @param comment The comment associated with the column.
   * @param nullable Whether the column value can be null.
   * @param autoIncrement Whether the column is an auto-increment column.
   * @param defaultValue The default value of the column.
   */
  private ColumnDTO(
      String name,
      Type dataType,
      String comment,
      boolean nullable,
      boolean autoIncrement,
      Expression defaultValue) {
    this.name = name;
    this.dataType = dataType;
    this.comment = comment;
    this.nullable = nullable;
    this.autoIncrement = autoIncrement;
    this.defaultValue = defaultValue == null ? Column.DEFAULT_VALUE_NOT_SET : defaultValue;
  }

  @Override
  public String name() {
    return name;
  }

  @Override
  public Type dataType() {
    return dataType;
  }

  @Override
  public String comment() {
    return comment;
  }

  @Override
  public boolean nullable() {
    return nullable;
  }

  @Override
  public boolean autoIncrement() {
    return autoIncrement;
  }

  @Override
  public Expression defaultValue() {
    return defaultValue;
  }

  @Override
  public Audit auditInfo() {
    return AuditDTO.builder().build();
  }

  /**
   * Creates a new Builder to build a Column DTO.
   *
   * @return A new Builder instance.
   */
  public static Builder builder() {
    return new Builder();
  }

  /**
   * Builder class for constructing ColumnDTO instances.
   *
   * @param <S> The type of the builder subclass.
   */
  public static class Builder<S extends Builder> {

    /** The name of the column. */
    protected String name;

    /** The data type of the column. */
    protected Type dataType;

    /** The comment associated with the column. */
    protected String comment;

    /** * Whether the column value can be null. */
    protected boolean nullable = true;

    /** Whether the column is an auto-increment column. */
    protected boolean autoIncrement = false;

    /** The default value of the column. */
    protected Expression defaultValue;

    /** Constructs a new Builder. */
    public Builder() {}

    /**
     * Sets the name of the column.
     *
     * @param name The name of the column.
     * @return The Builder instance.
     */
    public S withName(String name) {
      this.name = name;
      return (S) this;
    }

    /**
     * Sets the data type of the column.
     *
     * @param dataType The data type of the column.
     * @return The Builder instance.
     */
    public S withDataType(Type dataType) {
      this.dataType = dataType;
      return (S) this;
    }

    /**
     * Sets the comment associated with the column.
     *
     * @param comment The comment associated with the column.
     * @return The Builder instance.
     */
    public S withComment(String comment) {
      this.comment = comment;
      return (S) this;
    }

    /**
     * Sets whether the column value can be null.
     *
     * @param nullable Whether the column value can be null.
     * @return The Builder instance.
     */
    public S withNullable(boolean nullable) {
      this.nullable = nullable;
      return (S) this;
    }

    /**
     * Sets whether the column is an auto-increment column.
     *
     * @param autoIncrement Whether the column is an auto-increment column.
     * @return The Builder instance.
     */
    public S withAutoIncrement(boolean autoIncrement) {
      this.autoIncrement = autoIncrement;
      return (S) this;
    }

    /**
     * Sets the default value of the column.
     *
     * @param defaultValue The default value of the column.
     * @return The Builder instance.
     */
    public S withDefaultValue(Expression defaultValue) {
      this.defaultValue = defaultValue;
      return (S) this;
    }

    /**
     * Builds a Column DTO based on the provided builder parameters.
     *
     * @return A new ColumnDTO instance.
     * @throws NullPointerException If required, fields name and data type are not set.
     */
    public ColumnDTO build() {
      Preconditions.checkNotNull(name, "Column name cannot be null");
      Preconditions.checkNotNull(dataType, "Column data type cannot be null");
      return new ColumnDTO(name, dataType, comment, nullable, autoIncrement, defaultValue);
    }
  }

  /**
   * Validates the Column DTO.
   *
   * @throws IllegalArgumentException If some of the required fields are not set or if the column is
   *     non-nullable with a null default value, this method will throw an IllegalArgumentException.
   */
  public void validate() throws IllegalArgumentException {
    if (name() == null || name().isEmpty()) {
      throw new IllegalArgumentException("Column name cannot be null or empty.");
    }
    if (dataType() == null) {
      throw new IllegalArgumentException("Column data type cannot be null.");
    }
    if (!nullable()
        && defaultValue() instanceof LiteralDTO
        && ((LiteralDTO) defaultValue()).dataType().equals(Types.NullType.get())) {
      throw new IllegalArgumentException(
          "Column cannot be non-nullable with a null default value: " + name());
    }
  }
}
